src/privileged/vpnRecommender/api.js

```diff
@@ -0,0 +1,537 @@
+    }
+
+    if (variation === "streaming-hostname") {
+      this.registerStreamingHostnameTrigger();
+    }
+  }
+
+  registerCaptivePortalTrigger() {
+    const that = this;
+
+    const cpObserver = {
+      observe(subject, topic, data) {
+        if (topic !== "captive-portal-login") return;
+        log("captive-portal-login");
+        that.waitForConnection().then((result) => {
+          if (result === "success") that.tryShowNotification();
+        });
+      },
+    };
+
+    Services.obs.addObserver(cpObserver, "captive-portal-login");
+    this.addCleanUpFunction(() => {
+      Services.obs.removeObserver(cpObserver, "captive-portal-login");
+    });
+  }
+
+  registerHostnameTrigger(hostnameList) {
+    log("registering hostname trigger");
+
+    const that = this;
+
+    const progressListener = {
+      QueryInterface: ChromeUtils.generateQI(["nsIWebProgressListener",
+        "nsISupportsWeakReference"]),
+
+      onStateChange(webProgress, request, flag, status) {
+      },
+
+      onLocationChange(progress, request, uri) {
+        // This fires when the location bar changes; that is load event is confirmed
+        // or when the user switches tabs. If you use myListener for more than one tab/window,
+        // use progress.DOMWindow to obtain the tab/window which triggered the change.
+
+        try {
+          const hostname = uri.host;
+
+          const test_function = (ref_hostname) => {
+            const escaped_ref_hostname = ref_hostname.replace(".", "\\.").replace("-", "\\-");
+            const rx = new RegExp("^(.+\\.)*(" + escaped_ref_hostname + ")$");
+
+            return rx.test(hostname);
+          };
+
+          if (hostnameList.some(test_function)) {
+            that.tryShowNotification();
+          }
+
+        } catch (e) {
+
+        }
+      },
+    };
+
+    // current windows
+    const windowEnumerator = Services.wm.getEnumerator("navigator:browser");
+
+    while (windowEnumerator.hasMoreElements()) {
+      const window = windowEnumerator.getNext();
+
+      if (PrivateBrowsingUtils.isWindowPrivate(window)) continue; // ignore private windows
+
+      const winWeak = Cu.getWeakReference(window);
+
+      const onOpenWindow = function(e) {
+        winWeak.get().gBrowser.addProgressListener(progressListener);
+        winWeak.get().removeEventListener("load", onOpenWindow);
+        that.addCleanUpFunction(() => {
+          if (winWeak.get()) {
+            winWeak.get().gBrowser.removeProgressListener(progressListener);
+          }
+        });
+      };
+
+      if (winWeak.get().gBrowser) {
+        winWeak.get().gBrowser.addProgressListener(progressListener);
+        that.addCleanUpFunction(() => {
+          if (winWeak.get()) {
+            winWeak.get().gBrowser.removeProgressListener(progressListener);
+          }
+        });
+      } else {
+        winWeak.get().addEventListener("load", onOpenWindow, true);
+      }
+    }
+
+    // new windows
+    const windowListener = {
```

Comment:
Instead of doing this window management manually I would strongly advise you to copy over the EveryWindow.jsm module (https://github.com/mozilla/FastBlockShield/blob/master/src/privileged/trackers/EveryWindow.js).

EveryWindow has a super simple API, you're probably mostly interested in `registerCallback`, which allows you to register a function that runs on every new window and on all existing windows. The big advantage of using this is that it's battle-proven and abstracts things away nicely, so that your code gets smaller, you can rest easier and I have an easier time reviewing :)

Of course you couldn't know this existed, we should make it part of the shield-study utils, since most studies need this sort of functionality one way or the other.

Let me know if you have any questions!

Reply:
Thanks for pointing this out. No I'm using EveryWindow in various parts of the code. It cleaned up the code a lot.